[ADD] l10n_tr_nilvera_einvoice_extended: Adding nilvera einvoice exte…

alkh-odoo opened the pull request:
[ADD] l10n_tr_nilvera_einvoice_extended: Adding Nilvera E-Invoice Extension  
[ADD] l10n_tr_nilvera_einvoice_extended: Adding Tax Offices

---

### Description of the issue/feature this PR addresses:

We do not support Withholding VAT (KDV Tevkifatı), Exemption Invoices, and Export Invoices (İhracat Faturaları) with our current e-invoicing solution in Türkiye. Those invoice types are legally important and required for both domestic and international sales for businesses. In addition, the inclusion of the Tax office in our current XML we send to Nilvera, both for `<cac:AccountingSupplierParty>` and `<cac:AccountingCustomerParty>`, is required by law.

This PR adds an extension module for Nilvera E-Invoicing. The purpose of this module is to extend the existing E-Invoicing functionality to support additional invoice types and scenarios to be compliant with the TR legal requirements.

---

### Invoice Scenario:

**Basic**  
Invoice Type:
- Sales  
- Withholding  
- Tax Exempt  
- Registered for Export  

**Export**  
Invoice Type:
- Tax Exempt  

**Public**  
Invoice Type:
- Sales  
- Withholding  
- Tax Exempt  
- Registered for Export  

---

### Current behavior before PR:

We don't support a scenario-based system to comply with TR legal requirements.

---

### Desired behavior after PR is merged:

We support a scenario-based system to comply with TR legal requirements.

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

khfa-odoo commented on addons/l10n_tr_nilvera_einvoice_extended/models/account_edi_xml_ubl_tr.py:
```diff
@@ -0,0 +1,426 @@
+            taxes.filtered(lambda t: not t.l10n_tr_tax_withholding_code_id),
+        )
+        for tax in taxes.filtered(lambda t: t.l10n_tr_tax_withholding_code_id):
+            tax_type_code = tax.l10n_tr_tax_withholding_code_id.code or "0015"
```
if code is a required field, you don't need a fallback value

khfa-odoo commented on addons/l10n_tr_nilvera_einvoice_extended/models/account_edi_xml_ubl_tr.py:
```diff
@@ -0,0 +1,426 @@
+        )
+        for tax in taxes.filtered(lambda t: t.l10n_tr_tax_withholding_code_id):
+            tax_type_code = tax.l10n_tr_tax_withholding_code_id.code or "0015"
+            tax_scheme_name = (
+                tax.l10n_tr_tax_withholding_code_id.with_context(lang="tr_TR").name
+                or "Gerçek Usulde KDV"
+            )
```
no need for a fallback here as well

alkh-odoo commented on addons/l10n_tr_nilvera_einvoice_extended/models/account_edi_xml_ubl_tr.py:
```diff
@@ -0,0 +1,426 @@
+
+    def _get_partner_party_vals(self, partner, role):
+        if role == "tr_municipality":
+            return self._get_tr_municipality_vals(partner)
```
I’m overriding this function to include tr_municipality as a role. This override supports extensibility and aligns with how other localization flows override account.edi.xml.ubl_21. It also helps maintain code cohesion by keeping it consistent with standard localization patterns. Keeping it this way is beneficial, as we may need this role in other parts of the localization in the future.

alkh-odoo replied:
I understand the value of this approach, but I prefer to return a specific dictionary for the Municipality to prevent any extra tags from being passed. I believe this is a more foolproof solution against potential changes. The _get_tr_municipality_vals method already provides the exact details required to render the party tag for a municipality. If additional fields are needed, the method can easily be overridden.

alkh-odoo replied:
I have moved all the data from the partner to this function that returns a dict of them
    def _get_municipality_vals(self):
        return {
            "party_vals": {
                "party_identification_vals": {
                    "id_attrs": {"schemeID": "VKN"},
                    "id": "1460415308",
                },
                "postal_address_vals": {
                    "city_subdivision_name": "Ulus",
                    "city_name": "Ankara",
                    "country_vals": {
                        "name": "Türkiye",
                    },
                },
                "party_name_vals": [
                    {"name": "Gümrük ve Ticaret Bakanlığı Gümrükler Genel Müdürlüğü- Bilgi İşlem Dairesi Başkanlığı"}
                ],
                "party_tax_scheme_vals": [{"tax_scheme_vals": {"name": "Ulus"}}],
            },
        }

khfa-odoo commented on addons/l10n_tr_nilvera_einvoice_extended/models/res_partner.py:
```diff
@@ -0,0 +1,17 @@
+    )
+
+    def _compute_display_name(self):
+        super()._compute_display_name()
+        for partner in self.filtered(lambda r: r._context.get("show_vat")):
+            if partner._context.get("show_vat") and partner.l10n_tr_tax_office_id:
+                partner.display_name = (
+                    f"{partner.display_name}\n{partner.l10n_tr_tax_office_id.name}"
+                )
```
I would prefer if you avoid calling super and build display_name the way you want it.
```py
if not self.context:
   return super
# your logic
```

khfa-odoo commented on addons/l10n_tr_nilvera_einvoice_extended/models/account_move.py:
```diff
@@ -0,0 +1,92 @@
+
+    def _get_partner_l10n_tr_nilvera_customer_alias_name(self):
+        self.ensure_one()
+        if self.l10n_tr_is_export_invoice:
+            return "urn:mail:ihracatpk@gtb.gov.tr"
```
I think you forgot this @alkh-odoo

alkh-odoo commented on addons/l10n_tr_nilvera_einvoice_extended/models/account_edi_xml_ubl_tr.py:
```diff
@@ -0,0 +1,426 @@
+            res.append(
+                {
+                    "id": tax_type_code,
+                    "percent": tax.l10n_tr_tax_withholding_code_id.percentage or False,
```
I have added the withholding taxes, and the validation will be done in a separate task if required.

khfa-odoo commented:
@alkh-odoo, done from my side.

proose commented on addons/l10n_tr_nilvera_einvoice_extended/models/account_move.py:
```diff
@@ -0,0 +1,99 @@
+        store=True,
+        readonly=False,
+        string="Exemption Reason",
+        domain=lambda self: str(self._l10n_tr_exemption_code_domain()),
```
What is usually done in cases like this is to create a Many2many field `l10n_tr_available_exemption_code_ids` (One2many should work too I think) on the same model, with a compute that will fetch all ids that you want to show. That way, you can use here a `domain=[('id', 'in', l10n_tr_available_exemption_code_ids)]`. It should be more clear than using a Binary, thought I'm not sure what's best performance wise...

A bit like what you can find with `available_payment_method_ids` (https://github.com/odoo/enterprise/blob/dbf35bc09150e86f44698a8217cb68b0de479d7c/account_batch_payment/models/account_batch_payment.py#L60)

alkh-odoo replied:
Honestly, this approach will make us do a search that is not needed. In addition, having the types we want to be returned as a list to the domain is much clearer than searching for the records we want using their IDs. For me, the Binary field approach is very easy to extend. It is your call; if you believe that making it a m2m is the best approach here, then I will change it. Complexity-wise wise for me, I see using a binary field is much cleaner + less complications with the ORM cause we don't need to do an extra search.

proose commented on addons/l10n_tr_nilvera_einvoice_extended/models/account_move.py:
```diff
@@ -0,0 +1,99 @@
+    @api.depends(
+        "l10n_tr_gib_invoice_scenario", "l10n_tr_gib_invoice_type", "partner_id"
+    )
+    def _compute_l10n_tr_exemption_code_id(self):
+        self.l10n_tr_exemption_code_id = False
```
I'm not so sure about this compute with the default... Yes, the value is set to the default when creating a move, but a move could be created with any value through the code, so you could have a l10n_tr_gib_invoice_scenario different from its default, and still have the l10n_tr_gib_invoice_type to its default. Would that be ok ?

I'm wondering if we should not just remove the default on l10n_tr_exemption_code_id and let the user decide, or really automate the process by changing the compute to actually set some value when needed

proose commented on addons/l10n_tr_nilvera_einvoice_extended/models/account_move.py:
```diff
@@ -0,0 +1,99 @@
+    )
+
+    l10n_tr_exemption_code_domain_list = fields.Binary(
+        compute="_compute_l10n_tr_exemption_code_domain_list"
```
One line please, or trailing comma (if you really keep this field)

proose commented on addons/l10n_tr_nilvera_einvoice_extended/models/l10n_tr_res_tax_office.py:
```diff
@@ -0,0 +1,12 @@
+from odoo import models, fields
```
Order please

yace-odoo commented on addons/l10n_tr_nilvera_einvoice_extended/models/res_company.py:
```diff
@@ -0,0 +1,14 @@
+    )
+    l10n_tr_nilvera_export_alias = fields.Char(
+        string="Nilvera Export Alias",
+        default="urn:mail:ihracatpk@gtb.gov.tr",
```
Hello @proose, answering from the functional perspective

Export invoices has to be sent with very specific data 

- Alias is given by the government: `urn:mail:ihracatpk@gtb.gov.tr`
    - For other scenarios (Basic, Public), the Alias is unique to each user
- In the XML under `cac:AccountingCustomerParty` needs to have the Ministry of Trade & Customs address

these invoices are directly sent to the Ministry rather than the customer (as per the regulation)

The way it was implemented was to have the default value for Export but also give the customers option to change it under the settings, should the government decide it would need changing.

@alkh-odoo

yace-odoo replied:
Also wanted to mention, for this ministry info and also the various reasons we've added, on the XML they cannot be translated it has to be in Turkish. 

So what we've done is in Odoo you can use any language and we have EN translations for it, but when an XML is generated corresponding TR value is slapped to XML

proose replied:
> Not really, scenarios are invoice specific

Does it make sense to put the alias on the company then ? Shouldn't it be put on the invoice ? Or have multiple alias to use for the different cases ?

> Gümrük ve Ticaret Bakanlığı Gümrükler Genel Müdürlüğü- Bilgi İşlem Dairesi Başkanlığı

That is indeed the value in the `l10n_tr_res_partner_municipality` partner data. But then we should change the references to it by "ministry" instead of "municipality", right ?

> it's in settings hidden under debug=1

Normally it should still be written in a view file, with a `group_no_one` group to allow only debug-users to see it

Thanks for the info !

yace-odoo replied:
> Does it make sense to put the alias on the company then ? Shouldn't it be put on the invoice ? Or have multiple alias to use for the different cases ?

Currently it's under contacts, because it's an info you put only once and forget about it. Only case where you send it to a different alias is Export Scenario so our implementation is, when user selects the Export Scenario, alias becomes `urn:mail:ihracatpk@gtb.gov.tr`

> That is indeed the value in the `l10n_tr_res_partner_municipality` partner data. But then we should change the references to it by "ministry" instead of "municipality", right ?

True :sweat_smile: 

> Normally it should still be written in a view file, with a `group_no_one` group to allow only debug-users to see it

Got it! Thanks, it should be corrected

alkh-odoo replied:
Hello, I found that I had not included the field in the XML, so I have updated this part.

As for the l10n_tr_res_partner_municipality we removed it fully.

alkh-odoo commented on addons/l10n_tr_nilvera_einvoice/models/account_move_send.py:
```diff
@@ -46,20 +46,53 @@ def _get_alerts(self, moves, moves_data):
+        # Overriden in l10n_tr_nilvera_einvoice_extended to give error based on l10n_tax_office field
         if tr_einvoice_partners_missing_ref := moves.partner_id.filtered(
-            lambda p: p.l10n_tr_nilvera_customer_status == "einvoice" and not p.ref
+            lambda p: p.l10n_tr_nilvera_customer_status == "einvoice"
+            and not p.ref
+            and p.country_code == "TR"
```
Could you please share the format you’re using? As it will help clean out the code better from my end. I’m currently using Ruff Formatter, which automatically applies these indentations.

Thank you in advance!

proose commented on addons/l10n_tr_nilvera_einvoice_extended/views/l10n_tr_account_tax_code_views.xml:
```diff
@@ -0,0 +1,61 @@
+    </record>
+
+    <menuitem
+        id="menu_account_tax_code"
```
Mayyybe a tad too generic (maybe a "l10n_tr" or just a "tr" in there, just to be a bit more futureproof)

alkh-odoo replied:
100%

proose commented on addons/l10n_tr_nilvera_einvoice_extended/views/l10n_tr_res_tax_office_views.xml:
```diff
@@ -0,0 +1,54 @@
+    <record id="l10n_tr_res_tax_office_view_list" model="ir.ui.view">
+        <field name="name">l10n_tr.res.tax.office.list</field>
+        <field name="model">l10n_tr.res.tax.office</field>
+        <field name="mode">primary</field>
```
Same comment than in the other file (no use of primary in a non-inheriting view)

alkh-odoo replied:
100%

proose commented on addons/l10n_tr_nilvera_einvoice_extended/views/l10n_tr_res_tax_office_views.xml:
```diff
@@ -0,0 +1,54 @@
+    <record id="l10n_tr_res_tax_office_view_form" model="ir.ui.view">
+        <field name="name">l10n_tr.res.tax.office.form</field>
+        <field name="model">l10n_tr.res.tax.office</field>
+        <field name="mode">primary</field>
```
Same comment than in the other file (no use of primary in a non-inheriting view)

alkh-odoo commented on addons/l10n_tr_nilvera_einvoice_extended/models/l10n_tr_res_tax_office.py:
```diff
@@ -0,0 +1,12 @@
+
+
+class L10nTrAccountTaxCode(models.Model):
+    _name = "l10n_tr.res.tax.office"
```
Yeh makes sense, I should have named it, l10n_tr_nilvera_einvoice_extended.tax.office but it will be a very long name.

As for the res, I thought since I am adding it under the section of res.country, res.city as it functions similarly to them,
<img width="1181" height="140" alt="image" src="https://github.com/user-attachments/assets/89c75d21-8443-4e91-929f-f7b5f836394b" />

I initially wanted to make it as res.tax.office, but yeah I have updated the module name and all of its references to reflect on this.

proose commented on addons/l10n_tr_nilvera_einvoice_extended/__manifest__.py:
```diff
@@ -0,0 +1,35 @@
+    3.Configuration of withholding reasons and exemption reasons
+    4.Addition of Tax Offices.
+    """,
+    'depends': ['l10n_tr_nilvera_einvoice', 'contacts'],
```
I just saw the view for tax offices. I suppose it is for that. But that means forcing the install of another module just to add it to a specific menu ? Unless I missed something, maybe you could put the menuitem for tax offices in the configuration menu of accounting ?
You could even go further and create a Turkish localisation group menuitem (might be a bit overkill for 2 menuitems)

proose commented on addons/l10n_tr_nilvera_einvoice_extended/models/product_template.py:
```diff
@@ -0,0 +1,18 @@
+    _inherit = "product.template"
+
+    l10n_tr_gibp_number = fields.Char(
+        string="GIBP Number",
```
I read "GTIP Number" in the task, is it the same thing ?

proose commented on addons/l10n_tr_nilvera_einvoice_extended/views/product_views.xml:
```diff
@@ -0,0 +1,31 @@
+            <field name="arch" type="xml">
+                    <xpath expr="//field[@name='barcode']" position="after">
+                        <field name="l10n_tr_gibp_number"
+                               invisible="product_variant_count &gt; 1
```
Why not allowing to see/change it when multiple variants of the same products exist ? Wouldn't they most likely have the same number or not ?

proose commented on addons/l10n_tr_nilvera_einvoice_extended/views/res_partner_views.xml:
```diff
@@ -0,0 +1,15 @@
+        <record id="view_partner_form_l10n_tr_nilvera_extended" model="ir.ui.view">
+            <field name="name">view.partner.form.inherit.l10n.tr.nilvera.extended</field>
+            <field name="model">res.partner</field>
+            <field name="inherit_id" ref="base.view_partner_form" />
```
```suggestion
            <field name="inherit_id" ref="base.view_partner_form"/>
```

proose commented on addons/l10n_tr_nilvera_einvoice_extended/views/res_config_settings_views.xml:
```diff
@@ -0,0 +1,21 @@
+        <field name="arch" type="xml">
+            <xpath expr="//setting[@id='nilvera_settings']/div[hasclass('content-group')]" position="inside">
+                <div class="row" groups="base.group_no_one">
+                    <label string="Export Alias" for="l10n_tr_nilvera_api_key" class="col-lg-6 o_light_label" />
```
```suggestion
                    <label string="Export Alias" for="l10n_tr_nilvera_api_key" class="col-lg-6 o_light_label"/>
```

proose commented on addons/l10n_tr_nilvera_einvoice_extended/views/res_partner_views.xml:
```diff
@@ -0,0 +1,15 @@
+            <field name="model">res.partner</field>
+            <field name="inherit_id" ref="base.view_partner_form" />
+            <field name="arch" type="xml">
+                <xpath expr="//group/div[hasclass('o_address_format')]" position="after">
```
```suggestion
                <xpath expr="//group/field[@name='vat']" position="before">
```
Might be a bit more clear/secure, as we can already find several cases of `class="o_address_format"` inside a group (on the same model), but that might just be me being too safe for nothing

alkh-odoo replied:
<img width="1085" height="194" alt="image" src="https://github.com/user-attachments/assets/380b57cf-1bb8-4d95-b680-aee2ca612699" />

alkh-odoo replied:
I will xpath div/field[@name='state_id']/.. should always be after the address div like we need it to be.

alkh-odoo commented on addons/l10n_tr_nilvera_einvoice/models/account_move_send.py:
Yes, in the context of AccountMoveSend, we need to check the reference field, as that was the original solution designed for the tax office. I've moved the logic to the _get_tax_company_tax_office_alert method so it can be easily overridden by the new module. 

In the new module, where tax offices are present, we will override this method and implement the logic based on the tax office instead of relying on the reference field. We need to keep the original solution of the reference intact, as there may be cases where the extended module isn't installed. Since this is compliance-related, it's important to continue supporting the original logic based on the reference field.

alkh-odoo replied:
TLDR, the original solution relied on the reference I have overriden this part to make it such that we keep the original flow and edit it in the new module.

alkh-odoo commented on addons/l10n_tr_nilvera_einvoice/models/account_move_send.py:
```diff
@@ -46,20 +46,53 @@ def _get_alerts(self, moves, moves_data):
                 "action_text": _("View Partner(s)"),
-                "action": tr_einvoice_partners_missing_ref._get_records_action(name=_("Check reference on Partner(s)")
+                "action": tr_einvoice_partners_missing_ref._get_records_action(
+                    name=_("Check reference on Partner(s)")
```
Yeah it is because of the Ruff formatter, 🙁 I will update it to be on one line and add the optional comma

proose replied:
No worries, these are some nitpicks (but they don't take much time to do :smile: )

Weird that the Ruff formatter doesn't automatically add the trailing comma, my Ruff linter shows a warning on it

alkh-odoo replied:
Oh, I dug around and it turns out my toml files lint.select was missing "COM". Now I can see all the missing trailing commas 😊

alkh-odoo replied:
If you could share where I can find all the important lint selects, it would be very appreciated

proose replied:
Just going through all comments again:
> I will update it to be on one line and add the optional comma

When you are in a one line, no need to add the comma (you can, of course, but we don't enforce it). We just prefer to add the comma when the param is on its own line, because if someone needs to add a new param, he would need to add the comma and the git history of the previous param line would be altered.
Examples of correct calls:
`call_to_a_function(param1, param2)`
`call_to_a_function(param1, param2,)` (less used, no real need of the comma)
```python
call_to_a_function(
    param1,
    param2,
)
```

alkh-odoo commented on addons/l10n_tr_nilvera_einvoice/models/account_move.py:
```diff
@@ -269,6 +269,12 @@ def _l10n_tr_nilvera_einvoice_check_invalid_subscription_dates(self):
 
+    def _get_partner_l10n_tr_nilvera_customer_alias_name(self):
+        # This getter function is used to change the alias on send
+        # For any senario spesific changes for the alias
```
Yeah, I've updated it. Just to note, this getter function is used in l10n_tr_nilvera_einvoice_extended for the export flow, where the alias needs to be different from the customer alias.

alkh-odoo commented on addons/l10n_tr_nilvera_einvoice_extended/data/template/account.tax-tr.csv:
```diff
@@ -0,0 +1,271 @@
+"","","","","","","","","","","tax","invoice","tr391004","-SAL_20_TAX_9_10","",""
+"","","","","","","","","","","base","refund","","-SAL_20_BASE_9_10","",""
+"","","","","","","","","","","tax","refund","tr391004","+SAL_20_TAX_9_10","",""
+"tr_s_wh_20_MEBH","3","Renovation, Maintenance and Repair Services for Machinery, Equipment, Fixed Assets and Vehicles","-14%","WH 20% (9/10) - MEBH","-14.0","percent","","none","tr_tax_group_wh_18","base","invoice","","+SAL_20_BASE_9_10","Makine, Teçhizat, Demirbaş ve Taşıtlara Ait Tadil, Bakım, Onarım Hizmetleri","tax_excluded","l10n_tr_nilvera_einvoice_extended.l10n_tr_account_tax_code_603"
```
actually it is just an issue with the naming 
l10n_tr_nilvera_einvoice_extended_account_tax_code_603 is the withholding reason, and it has 0.7, so 70% yeh I think I missed this one while naming them. since

"","","","","","","","","","","tax","invoice","tr391005","-PUR_20_TAX_7_10","",""
"","","","","","","","","","","base","refund","","-PUR_20_BASE_7_10","",""
"","","","","","","","","","","tax","refund","tr391005","+PUR_20_TAX_7_10","",""
so it should be 7/10

alkh-odoo commented on addons/l10n_tr_nilvera_einvoice_extended/data/template/account.tax-tr.csv:
```diff
@@ -0,0 +1,271 @@
+"","","","","","","","","","","tax","invoice","tr391005","-PUR_20_TAX_9_10","",""
+"","","","","","","","","","","base","refund","","-PUR_20_BASE_9_10","",""
+"","","","","","","","","","","tax","refund","tr391005","+PUR_20_TAX_9_10","",""
+"tr_p_wh_20_MEBH","3","Purchasing Renovation, Maintenance and Repair Services for Machinery, Equipment, Fixed Assets and Vehicles","-14%","WH 20% (9/10) - MEBH","-14.0","percent","","none","tr_tax_group_wh_18","base","invoice","","+PUR_20_BASE_9_10","Makine, Teçhizat, Demirbaş ve Taşıtlara Ait Tadil, Bakım, Onarım Hizmetleri","tax_excluded","l10n_tr_nilvera_einvoice_extended.l10n_tr_account_tax_c ...
```
Similar to the one above, I have updated them to be 7/10 so -14% on withholding tax

alkh-odoo commented on addons/l10n_tr_nilvera_einvoice_extended/data/demo/demo_company.xml:
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo>
+    <record id="base.demo_company_tr" model="res.company" forcecreate="1">
+        <field name="l10n_tr_nilvera_api_key">urn:mail:ihracatpk@gtb.gov.tr</field>
```
Yes, this was totally missed

alkh-odoo commented on addons/l10n_tr_nilvera_einvoice_extended/data/l10n_tr_account_tax_code_data.xml:
```diff
@@ -0,0 +1,240 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo>
+    <data noupdate="0">
```
Yes, it should be noupdate="1". I put it as 0 just for testing forgot to change it 😭

alkh-odoo commented on addons/l10n_tr_nilvera_einvoice_extended/data/l10n_tr_account_tax_code_data.xml:
```diff
@@ -0,0 +1,240 @@
+<odoo>
+    <data noupdate="0">
+        <!-- Tax Exception Codes (Exempted 0% Tax) -->
+        <record id="l10n_tr_account_tax_code_212" model="l10n_tr.account.tax.code">
```
So for the Tax codes that are withholding codes, such as the 6XX codes, we have to have taxes for them on the item cause they are per item ("account.move.line"). As for the other tax codes, they are per invoice, so we added new fields on the invoice to dynamically pull the required codes based on the scenario and scenario type.

You can see the function here:
https://github.com/odoo/odoo/pull/213766/files#diff-2c3afb73c5e80203f1568994089fc237620a1275f58b9917e6f0b1d0ea5b587aR73

proose commented on addons/l10n_tr_nilvera_einvoice_extended/data/l10n_tr_account_tax_code_data.xml:
```diff
@@ -0,0 +1,240 @@
+            <field name="code_type">withholding</field>
+        </record>
+        <record id="l10n_tr_account_tax_code_606" model="l10n_tr.account.tax.code">
+            <field name="name">Labor Supply Services</field>
```
Thanks for highlighting this part for me, I missed that :smile:

alkh-odoo commented on addons/l10n_tr_nilvera_einvoice_extended/data/res_partner_data.xml:
Ok, so yes, that partner is inactive, so the users don't delete it by mistake. My goal was to make it dynamic and easy to change from the user's end, so that if there is an issue with the values we are sending, they can manually update the partner record. 

A dictionary is a very minimal, easy, and quick solution. Honestly, I will just do it for now. will just return a dictionary with the values as you said, too many functions that are only used once.

alkh-odoo replied:
Done.

alkh-odoo replied:
Very valid, I understand your point of view on this and I agree with it.

alkh-odoo commented on addons/l10n_tr_nilvera_einvoice_extended/models/l10n_tr_res_tax_office.py:
```diff
@@ -0,0 +1,12 @@
+from odoo import models, fields
+
+
+class L10nTrAccountTaxCode(models.Model):
```
Yes, updated with new name

alkh-odoo commented on addons/l10n_tr_nilvera_einvoice_extended/models/product_product.py:
```diff
@@ -0,0 +1,16 @@
+from odoo import api, fields, models, _
```
Yeh, after I updated the Ruff linter, I can see all those issues, but yeh I kinda missed this one.

alkh-odoo commented on addons/l10n_tr_nilvera_einvoice_extended/models/account_move_send.py:
```diff
@@ -0,0 +1,82 @@
+                "level": "warning",
+                "action_text": _("View Product(s)"),
+                "action": non_eligible_tr_products._get_records_action(
+                    name=_("Check Products")
```
100%

alkh-odoo commented on addons/l10n_tr_nilvera_einvoice_extended/models/account_move_send.py:
```diff
@@ -0,0 +1,82 @@
+    def _get_alerts(self, moves, moves_data):
+        # EXTENDS 'account'
+        alerts = super()._get_alerts(moves, moves_data)
+        tr_moves = self._get_l10n_tr_moves(moves)
```
Yeah, I’m a bit obsessed with making things easily accessible through getters, since they’re easier to override—unlike having the logic directly inside the function, where you’d need to filter the moves and pass them to super. But I get your point. It’s only used once, so there’s no real need to turn it into a generic function. will change that.

alkh-odoo replied:
Btw for this part, I am changing it to be
        tr_moves = self.filtered(lambda m: m.country_code == "TR")
        alerts = super()._get_alerts(tr_moves, moves_data)
cause we only want this warning to be on moves with code TR

alkh-odoo commented on addons/l10n_tr_nilvera_einvoice_extended/models/account_move_send.py:
```diff
@@ -0,0 +1,82 @@
+        tr_moves = self._get_l10n_tr_moves(moves)
+
+        # Warning alert if product is missing GIBP Number
+        if non_eligible_tr_products := self._get_l10n_tr_move_non_eligible_tr_products(
```
Oh btw for this one it was too long, I think I got a runbot warning saying to move it to a function

alkh-odoo commented on addons/l10n_tr_nilvera_einvoice_extended/models/account_move.py:
100% I have changed it, it was just during development, I just wanted to have them visually separated

alkh-odoo commented:
> Here the rest of my review, except for the tests. Thanks for the work. It will need some more tests to check the additional flows you added with this module (different taxes combinations, partners, tax offices, ...) as the existing ones don't.
> 
> PS: when commenting about one lines, I realized you probably followed the usual PEP8 linter, so I can't force you to make longer lines (FYI: the Odoo doc says you can ignore the PEP8 limit for a line), but please then add the trailing commas when needed (IMO, the one lines would be better).
> 
> Thanks again

Thank you very much for the hard work you put into this review. I have removed all the extra lines plus I added the trailing commas. I do what is best for the cleanliness of the code, as I want our localizations to be as clean as standard modules. I do appreciate every single comment about the commas, the extra lines, and so on. It is viable stuff that I will share with my colleagues.

alkh-odoo commented on addons/l10n_tr_nilvera_einvoice_extended/models/res_config_settings.py:
```diff
@@ -0,0 +1,11 @@
+class ResConfigSettings(models.TransientModel):
+    _inherit = 'res.config.settings'
+
+    l10n_tr_nilvera_export_alias = fields.Char(
```
This field is the one in res_company we are just pulling it to the config page.

alkh-odoo commented on addons/l10n_tr_nilvera_einvoice_extended/models/res_partner.py:
```diff
@@ -0,0 +1,20 @@
+
+    @api.depends('l10n_tr_tax_office_id')
+    def _compute_display_name(self):
+        if not self._context.get("show_address"):
+            return super()._compute_display_name()
+        super()._compute_display_name()
```
So this context is used in the view to turn the m2o res_partner field into the partner name with a full address,

It is mainly for updating the Turkish addresses to include the tax office name in any place that has this show_address set to true like account.move.form

https://github.com/odoo/odoo/blob/c10e44ceb6d0086cf659fdba851014eb3605ecfe/addons/account/views/account_move_views.xml#L978

<img width="257" height="123" alt="image" src="https://github.com/user-attachments/assets/4396c746-9601-413c-aa72-f52a175ea691" />
<img width="787" height="350" alt="image" src="https://github.com/user-attachments/assets/11759c80-a0d8-4eb4-9d79-5e129e89fab5" />

alkh-odoo commented on addons/l10n_tr_nilvera_einvoice_extended/security/ir.model.access.csv:
yeh, I have removed " and changed the . with _

alkh-odoo commented on addons/l10n_tr_nilvera_einvoice_extended/security/ir.model.access.csv:
```diff
@@ -0,0 +1,8 @@
+access_l10n_tr_account_tax_code,l10n_tr.account.tax.code,model_l10n_tr_account_tax_code,account.group_account_basic,1,1,1,1
+"access_l10n_tr_res_tax_office_public","l10n_tr_res_tax_office group_public_user_all","model_l10n_tr_res_tax_office",base.group_public,1,0,0,0
+"access_l10n_tr_res_tax_office_portal","l10n_tr_res_tax_office group_protal_user_all","model_l10n_tr_res_tax_office",base.group_portal,1,0,0,0
+"access_l10n_tr_res_tax_office_employee","l10n_tr_res_tax_office group_user_all","model_l10n_tr_res_tax_office",base.group_user,1,0,0,0
+"access_l10n_tr_res_tax_office_group_user","l10n_tr_res_tax_office group_user","model_l10n_tr_res_tax_office","base.group_partner_manager",1,1,1,1
```
For this, I kept it consistent with res. country 
https://github.com/odoo/odoo/blob/dc25929f3b7e1c29d540fc84b277c120d0bea4e6/odoo/addons/base/security/ir.model.access.csv#L44

I was as confused as you when I read the., I thought since it was the base module I had to name our access rights similar to what is in res.country, but yeah I am more than happy to give them better names that make sense to us.

alkh-odoo commented on addons/l10n_tr_nilvera_einvoice_extended/security/ir.model.access.csv:
```diff
@@ -0,0 +1,8 @@
+"access_l10n_tr_res_tax_office_portal","l10n_tr_res_tax_office group_protal_user_all","model_l10n_tr_res_tax_office",base.group_portal,1,0,0,0
+"access_l10n_tr_res_tax_office_employee","l10n_tr_res_tax_office group_user_all","model_l10n_tr_res_tax_office",base.group_user,1,0,0,0
+"access_l10n_tr_res_tax_office_group_user","l10n_tr_res_tax_office group_user","model_l10n_tr_res_tax_office","base.group_partner_manager",1,1,1,1
+"access_l10n_tr_res_tax_office_group_system","l10n_tr_res_tax_office_group_system","model_l10n_tr_res_tax_office","base.group_system",1,1,1,1
```
Oh I did not test without it, since I took the access rights of res.company as is.

alkh-odoo commented on addons/l10n_tr_nilvera_einvoice_extended/data/ubl_tr_templates.xml:
It won't work any other way, sorry. The accounts chart is hard-coded to access data/templates for account chart-related data.

This is the line:
https://github.com/odoo/odoo/blob/dc25929f3b7e1c29d540fc84b277c120d0bea4e6/addons/account/models/chart_template.py#L1263

alkh-odoo commented on addons/l10n_tr_nilvera_einvoice_extended/data/ubl_tr_templates.xml:
```diff
@@ -0,0 +1,57 @@
+                <cac:DeliveryAddress>
+                    <t t-set="delivery_address_vals" t-value="vals.get('delivery_address_vals', {})"/>
+                    <cbc:StreetName t-out="delivery_address_vals.get('street')"/>
+				    <cbc:CitySubdivisionName t-out="delivery_address_vals.get('city')"/>
```
Should be clean now?

proose commented:
Don't hesitate to re-request my review once you push all your changes :wink:

alkh-odoo commented on addons/l10n_tr_nilvera_einvoice_extended/models/account_edi_xml_ubl_tr.py:
```diff
@@ -0,0 +1,389 @@
+        if (
+            invoice.l10n_tr_gib_invoice_type == "TEVKIFAT"
+            and tax.l10n_tr_tax_withholding_code_id
+        ):
```
100% since we don't have invoice access, the only place is to set it here, or we can move them to a new function and set it there.

Will add the comment

alkh-odoo commented on addons/l10n_tr_nilvera_einvoice_extended/models/account_edi_xml_ubl_tr.py:
```diff
@@ -0,0 +1,389 @@
+        tax_totals_vals = super(
+            AccountEdiXmlUblTr, self.with_context(skip_code=True)
+        )._get_invoice_line_tax_totals_vals_list(line, taxes_vals)
+        return copy.deepcopy(tax_totals_vals)
```
Ok, so there is an issue with this function, and the issue is not easy to replicate, but what is happening is that _export_invoice_vals calls first the _get_invoice_tax_totals_vals_list
. This function creates an object and returns it to _export_invoice_vals. Now the issue is here in account_edi_ubl_cii
 https://github.com/odoo/odoo/blob/995a7072cb3315fc03544b281b1ed5ca4e81e901/addons/account_edi_ubl_cii/models/account_edi_xml_ubl_20.py#L504

_get_invoice_line_tax_totals_vals_list calls _get_invoice_tax_totals_vals_list, which returns the same object to the lines. So what ends up happening is that when I update the values inside _get_invoice_line_vals, tax_vals points to the same object. When I edit them, the changes also affect the invoice tax values.

Detailed explanation:
So  _export_invoice_vals function first of all calls  _get_invoice_line_tax_totals_vals_list while calculating the invoice_line_vals_list.
<img width="1061" height="156" alt="image" src="https://github.com/user-attachments/assets/384e3da2-c9c7-4426-b540-158e53c91301" />

This creates a new object of type list for the invoice line values

The issue is that _get_invoice_line_tax_totals_vals_list calls the _get_invoice_tax_totals_vals_list of the invoice which is done to allow us to add more data to the total vals if needed, but _get_invoice_tax_totals_vals_list is creating only one object when it retruns if we end up doing any edits on the object to add more info on the lines it will end up changing the data for the object on the total invoice.
<img width="1079" height="115" alt="image" src="https://github.com/user-attachments/assets/1000c82b-e410-4e30-898a-0196ef6bbff3" />
since tax_total_vals gets the same object when calling the _get_invoice_tax_totals_vals_list
<img width="616" height="85" alt="image" src="https://github.com/user-attachments/assets/b39f2bf0-f8f5-4d60-8ad5-14a603380fb5" />

It kind of happens randomly from time to time, and I haven’t found a consistent way to reproduce it. The solution is to deep copy so we 100% of the time have a new object in memory before updating any values of the object.

alkh-odoo replied:
This could be something we can push in a separate commit for account_edi_ubl_cii IMP

alkh-odoo commented on addons/l10n_tr_nilvera_einvoice_extended/models/account_edi_xml_ubl_tr.py:
```diff
@@ -0,0 +1,389 @@
+        for vals in vals_list:
+            if partner.l10n_tr_tax_office_id:
+                vals["tax_scheme_vals"].update(
+                    {"id": "", "name": partner.l10n_tr_tax_office_id.name}
```
Because they are checking for the validity of the UBL 1.2

alkh-odoo commented on addons/l10n_tr_nilvera_einvoice_extended/__init__.py:
```diff
@@ -0,0 +1,17 @@
+    """Existing companies that have the Turkish Chart of Accounts set"""
+    template_codes = ["tr"]
+    tr_companies = env["res.company"].search(
+        [("chart_template", "in", template_codes), ("parent_id", "=", False)]
```
Removed

alkh-odoo commented on addons/l10n_tr_nilvera_einvoice_extended/views/account_move_views.xml:
```diff
@@ -0,0 +1,45 @@
+                        readonly="state in ['cancel', 'posted']"/>
+                <div class="o_td_label" invisible="l10n_tr_is_export_invoice or country_code != 'TR' or move_type != 'out_invoice' or l10n_tr_nilvera_customer_status not in ['einvoice', 'earchive']">
+                     <label for="l10n_tr_gib_invoice_scenario" invisible="l10n_tr_nilvera_customer_status not in ['einvoice']" />
+                     <label string="Invoice Type" for="l10n_tr_gib_invoice_type" invisible="l10n_tr_nilvera_customer_status not in ['earchive']"/>
```
I want to have dynamic naming for the fields, depending on the customer status, to be able to show and hide the label because both of the fields are on the same line.

As for the string, if I add it in python the Runbot would give a warning that the label is already used by another field which is the standard invoice_type

alkh-odoo commented on addons/l10n_tr_nilvera_einvoice_extended/views/account_move_views.xml:
Thank you for pointing this out. I have fixed them. I made everything into one line to make it much more compact and neat.

alkh-odoo commented:
@proose I have made all the changes and left comments on some

alkh-odoo commented on addons/l10n_tr_nilvera_einvoice/models/account_move_send.py:
```diff
@@ -46,18 +46,45 @@ def _get_alerts(self, moves, moves_data):
-            lambda p: p.l10n_tr_nilvera_customer_status == "einvoice" and not p.ref
+            lambda p: p.l10n_tr_nilvera_customer_status == "einvoice"
+            and not p.ref
+            and p.country_code == "TR",
```
Hahaha old habits die slow. Just changed it to a one-liner.

alkh-odoo commented on addons/l10n_tr_nilvera_einvoice_extended/__manifest__.py:
```diff
@@ -0,0 +1,34 @@
+    'depends': ['l10n_tr_nilvera_einvoice', 'contacts'],
+    'data': [
+        'security/ir.model.access.csv',
+        # 'demo/demo_company.xml',
```
Yes :(
![one-piece-chopper](https://github.com/user-attachments/assets/39954e3f-8b04-4722-b30c-f788fe826492)

proose commented on addons/l10n_tr_nilvera_einvoice_extended/models/account_move_line.py:
```diff
@@ -0,0 +1,25 @@
+from odoo.exceptions import ValidationError
+
+from odoo import _, api, fields, models
```
Weird, the linter does the opposite on my side

alkh-odoo commented on addons/l10n_tr_nilvera_einvoice_extended/models/account_edi_xml_ubl_tr.py:
```diff
@@ -0,0 +1,324 @@
+        grouping_key = super()._get_tax_grouping_key(base_line, tax_data)
+        invoice = base_line["record"].move_id
+        tax = tax_data["tax"]
+        grouping_key["_tax_category_vals_"]["l10n_tr_tax_exemption_reason"] = tax.l10n_tr_tax_withholding_code_id.id
```
True will do that.

alkh-odoo commented on addons/l10n_tr_nilvera_einvoice_extended/models/account_edi_xml_ubl_tr.py:
```diff
@@ -0,0 +1,324 @@
+
+        Used to fill the cac:InvoiceLine/cac:Item node in UBL TR XML export.
+
+        :param line:        An invoice line.
+        :return:            A dictionary with delivery information.
```
Fixed

alkh-odoo commented on addons/l10n_tr_nilvera_einvoice_extended/models/account_move_send.py:
```diff
@@ -0,0 +1,55 @@
+        # OVERRIDES l10n_tr_nilvera_einvoice.
+        if tr_einvoice_partners_missing_ref := moves.partner_id.filtered(lambda p: p.l10n_tr_nilvera_customer_status == "einvoice" and not p.l10n_tr_tax_office_id):
+            return {
+                "message": _("The following TR Partner(s) are missing the Tax Office Field."),
```
Sure, the second one seems much better.

alkh-odoo commented on addons/l10n_tr_nilvera_einvoice_extended/models/product_product.py:
```diff
@@ -0,0 +1,15 @@
+from odoo.exceptions import ValidationError
+
+from odoo import _, api, fields, models
```
Yes, I have, maybe the version of Ruff Linter that I am using is wrong?

alkh-odoo commented on addons/l10n_tr_nilvera_einvoice_extended/models/account_edi_xml_ubl_tr.py:
```diff
@@ -0,0 +1,324 @@
+
+        return vals
+
+    def _get_municipality_vals(self):
```
sure.

alkh-odoo commented on addons/l10n_tr_nilvera_einvoice_extended/tests/test_xml_ubl_tr.py:
```diff
@@ -0,0 +1,179 @@
+            'email': 'info@company.trexample.com',
+            'phone': '+90 501 234 56 78',
+            'l10n_tr_tax_office_id': cls.env.ref("l10n_tr_nilvera_einvoice_extended.l10n_tr_nilvera_einvoice_extended_tax_office_1009").id,
+            'bank_ids': [(0, 0, {'acc_number': 'TR0123456789'})],
```
Yeah, this is from the old tests done previously. Let me update it on both here and the other module, which is l10n_tr_nilvera_einvoice

alkh-odoo commented on addons/l10n_tr_nilvera_einvoice_extended/tests/test_xml_ubl_tr.py:
```diff
@@ -0,0 +1,179 @@
+        cls.tax_20 = cls.env['account.chart.template'].ref('tr_s_20')
+        cls.tax_20_withholding = cls.env['account.chart.template'].ref('tr_s_vat_wh_20_OGH')
+        # Registered for Export Reason
+        cls.reason_701 = cls.env.ref("l10n_tr_nilvera_einvoice_extended.l10n_tr_nilvera_einvoice_extended_account_tax_code_701")
+
+        # Tax Exemption Reason
+        cls.reason_212 = cls.env.ref("l10n_tr_nilvera_einvoice_extended.l10n_tr_nilvera_einvoice_extended_account_tax_code_212")
+
+        # Export Reason
+        cls.reason_301 = cls.env.ref("l10n_tr_nilvera_einvoice_extended.l10n_tr_nilvera_einvoice_extended_account_tax_code_301")
```
True

alkh-odoo commented on addons/l10n_tr_nilvera_einvoice_extended/views/res_company_views.xml:
```diff
@@ -0,0 +1,13 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo>
+    <record id="view_company_form_inherit_l10n_nz" model="ir.ui.view">
+        <field name="name">res.company.form.inherit.l10n_nz</field>
```
GG

alkh-odoo commented on addons/l10n_tr_nilvera_einvoice_extended/views/account_move_views.xml:
```diff
@@ -0,0 +1,33 @@
+                </div>
+                <div class="d-flex" invisible="l10n_tr_is_export_invoice or country_code != 'TR' or move_type != 'out_invoice' or l10n_tr_nilvera_customer_status not in ['einvoice', 'earchive']">
+                    <field name="l10n_tr_gib_invoice_scenario" invisible="l10n_tr_nilvera_customer_status not in ['einvoice']" readonly="state in ['cancel', 'posted']"/>
+                    <span class="o_form_label mx-3 oe_edit_only" invisible="l10n_tr_nilvera_customer_status not in ['einvoice']">is</span>
```
https://github.com/odoo/odoo/blob/3966753eb5a8534c8b5b8a16e626f5250c9013cf/addons/account/views/account_move_views.xml#L1050

Did it similarly to this 'or', which is a field above the new field. I wanted to match the look and feel of the due_date field, so I took classes from it.

rasardtechnology commented:
Hello everyone,

Do you have any time plan for the this PR is going to live? We are waiting this feature suddenly with our customers :)

proose reviewed (changes requested):
Thanks for those changes !
You forgot some comments though :D
Also the tests don't work, please make sure they run before pushing/requesting a new review.
I will already put it in our final reviewers' list because it seems urgent, but please make the changes asap so they don't have to go over it again :D

malb-odoo commented on addons/l10n_tr_nilvera_einvoice_extended/models/product_product.py:
```diff
@@ -0,0 +1,15 @@
+from odoo.exceptions import ValidationError
+
+from odoo import _, api, fields, models
```
https://www.odoo.com/documentation/17.0/contributing/development/coding_guidelines.html#imports
```suggestion
from odoo import _, api, fields, models

from odoo.exceptions import ValidationError
```

alkh-odoo commented on addons/l10n_tr_nilvera_einvoice_extended/models/product_template.py:
```diff
@@ -0,0 +1,18 @@
+    def _compute_l10n_tr_ctsp_number(self):
+        self._compute_template_field_from_variant_field("l10n_tr_ctsp_number")
+
+    def _set_l10n_tr_ctsp_number(self):
```
I want to keep it consistent with product.template/product.product shared fields as they all do _set_

malb-odoo commented:
@ThomasBeckers or @aboo-odoo Can we have the security review pleaaaase

malb-odoo commented:
Since i'm off tomorrow and next week, i will give the delegate to @alkh-odoo. For me the pr is good functionally and technically so once the security review is done, you will have to ask runbot for an exception and then you will be able to r+ it

@robodoo delegate=alkh-odoo

alkh-odoo commented on addons/l10n_tr_nilvera_einvoice_extended/tests/test_xml_ubl_tr.py:
```diff
@@ -0,0 +1,180 @@
+            'zip': '06810',
+            'city': 'İç Anadolu Bölgesi',
+            'state_id': cls.env.ref('base.state_tr_81').id,
+            'country_id': cls.env.ref('base.tr').id,
+            'email': 'info@company.trexample.com',
+            'phone': '+90 501 234 56 78',
+            'l10n_tr_tax_office_id': cls.env.ref("l10n_tr_nilvera_einvoice_extended.l10n_tr_nilvera_einvoice_extended_tax_office_1009").id,
+            'bank_ids': [Command.create({'acc_number': 'TR0123456789'})],
+        })
+
+        cls.einvoice_partner = cls.env['res.partner'].create({
+            'name': 'einvoice_partner',
+            'vat': '1729171602',
+            'street': 'Gökhane Sokak No:1',
+            'zip': '06934',
+            'city': 'Sincan/Ankara',
+            'state_id': cls.env.ref('base.state_tr_06').id,
+            'country_id': cls.env.ref('base.tr').id,
+            'email': 'info@tr_partner.com',
+            'phone': '+90 509 876 54 32',
+            'bank_ids': [Command.create({'acc_number': 'TR9876543210'})],
+            'invoice_edi_format': 'ubl_tr',
+            'l10n_tr_tax_office_id': cls.env.ref("l10n_tr_nilvera_einvoice_extended.l10n_tr_nilvera_einvoice_extended_tax_office_1009").id,
+            'l10n_tr_nilvera_customer_status': 'einvoice',
+        })
+
+        cls.earchive_partner = cls.env['res.partner'].create({
+            'name': 'earchive_partner',
+            'vat': '17291716060',
+            'street': 'Gökhane Sokak No:1',
+            'zip': '06934',
+            'city': 'Sincan/Ankara',
+            'state_id': cls.env.ref('base.state_tr_06').id,
+            'country_id': cls.env.ref('base.tr').id,
+            'email': 'info@tr_partner.com',
+            'phone': '+90 509 876 54 32',
+            'bank_ids': [Command.create({'acc_number': 'TR9876543210'})],
+            'invoice_edi_format': 'ubl_tr',
+            'l10n_tr_nilvera_customer_status': 'earchive',
+        })
```
Will do that!

alkh-odoo commented on addons/l10n_tr_nilvera_einvoice_extended/models/account_move_send.py:
```diff
@@ -0,0 +1,65 @@
+
+    def _get_l10n_tr_tax_partner_address_alert(self, moves, moves_data):
+        # EXTENDS l10n_tr_nilvera_einvoice.
+        return super()._get_l10n_tr_tax_partner_address_alert(moves.filtered(lambda move: not move.l10n_tr_is_export_invoice), moves_data)
```
There shouldn't be any case where there is a mismatch between data and moves.

aboo-odoo commented on addons/l10n_tr_nilvera_einvoice_extended/models/account_move.py:
```diff
@@ -0,0 +1,93 @@
+
+    def _l10n_tr_exemption_code_domain(self):
+        return [
+            ("code_type", "in", unquote("l10n_tr_exemption_code_domain_list") or []),
```
Why `unquote` ? The strings are not translated from what I can read ? Does it not work without ?

alkh-odoo replied:
Since the field values are not available at domain construction, loading them directly will not work for building a domain. Instead, using unquote ensures the field is evaluated when the data is loaded, allowing us to define a truly dynamic domain. Additionally, `l10n_tr_exemption_code_domain_list` is compute-dependent, which means the domain will dynamically update whenever the depends is triggered.

Examples of what I am trying to do in other parts of standard:
https://github.com/odoo/odoo/blob/9a9b162f4ea4fb34625eadaabae5aa1b1801aa48/addons/sale_timesheet/models/hr_timesheet.py#L30

Or

https://github.com/odoo/odoo/blob/9a9b162f4ea4fb34625eadaabae5aa1b1801aa48/addons/sale_project/models/project_project.py#L22

aboo-odoo replied:
As discussed, the best way to achieve your goal is to add the domain in the XML so that the domain_list is computed dynamically without the need to trick the system with `unquote` :+1:

alialfie commented:
As discussed, skipping all the tests in `l10n_tr_nilvera_einvoice` isn't good since it leaved dead code, a PR https://github.com/odoo/odoo/pull/229724 has been made to remove that part up to 18.3. The FW port for this PR in 18.4 is not yet merged, so this change should be done there before merging.
